Addition of new assignments and images

pranavk-2003 opened the pull request:
Addition of module 6 and 7 with images.

Module 6 contains 2 assignments related to line following tasks. the first assignment is introduction to the line sensor and its working, the second assignment implementing a relay controller for the line following tasks. Assignment is passed when two checkpoints are passed by the robot.

Module 7 contains one assignment which give the overview of P,PI and PID controllers and need the student to write a code for PID controller for the line follower robot.

ShulcN commented on lessons/module_7/PID.md:
```diff
@@ -0,0 +1,201 @@
+- \( e(t) \) is the error (desired value - actual value).
+
+### Block Diagram
+![P Controller](https://github.com/pranavk-2003/line-robot-curriculum/blob/main/images/module_7/p.png)
```
replace `pranavk-2003` to `autolab-fi`

Also add `?raw=True` in the end of the link, like below

`![image](https://github.com/autolab-fi/line-robot-curriculum/blob/main/images/module_2/headlights_2.png?raw=True)`

ShulcN commented on lessons/module_6/basic_line_follower.md:
```diff
@@ -0,0 +1,108 @@
+### **IR Sensor Configuration**  
+The **Octoliner** sensor array consists of **8 sensors**, each providing an analog value that indicates the intensity of reflected IR light.  
+
+![IR Sensor Logic](https://github.com/pranavk-2003/line-robot-curriculum/blob/main/images/module_6/IR_sensor_array.png)  
```
replace `pranavk-2003` to `autolab-fi`

Also add `?raw=True` in the end of the link, like below

`![image](https://github.com/autolab-fi/line-robot-curriculum/blob/main/images/module_2/headlights_2.png?raw=True)`

ShulcN commented on lessons/module_6/line_senosr_intro.md:
```diff
@@ -0,0 +1,76 @@
+
+For better visualization of the logic flow, refer to the flowchart below:
+
+![Flowchart](https://github.com/pranavk-2003/line-robot-curriculum/blob/main/images/module_6/FC_module_6.png)
```
replace `pranavk-2003` to `autolab-fi`

Also add `?raw=True` in the end of the link, like below

`![image](https://github.com/autolab-fi/line-robot-curriculum/blob/main/images/module_2/headlights_2.png?raw=True)`

ShulcN commented on lessons/module_7/PID.md:
```diff
@@ -0,0 +1,201 @@
+- The integral term accumulates past errors to eliminate steady-state error.
+
+### Block Diagram
+![PI Controller](https://github.com/pranavk-2003/line-robot-curriculum/blob/main/images/module_7/pi.png)
```
replace `pranavk-2003` to `autolab-fi`

Also add `?raw=True` in the end of the link, like below

`![image](https://github.com/autolab-fi/line-robot-curriculum/blob/main/images/module_2/headlights_2.png?raw=True)`

ShulcN commented on lessons/module_7/PID.md:
```diff
@@ -0,0 +1,201 @@
+- The derivative term (\( de(t)/dt \)) reduces rapid changes and dampens oscillations.  
+
+### Block Diagram  
+![PID Controller](https://github.com/pranavk-2003/line-robot-curriculum/blob/main/images/module_7/pid_f.png)  
```
replace `pranavk-2003` to `autolab-fi`

Also add `?raw=True` in the end of the link, like below

`![image](https://github.com/autolab-fi/line-robot-curriculum/blob/main/images/module_2/headlights_2.png?raw=True)`

ShulcN reviewed (commented):
1. The lesson numbers are incorrect, but I am not sure how important this is
2. The lesson about the PID regulator is quite lengthy, but it might be acceptable as a temporary lesson.


@spivakdm  what do you think about it?

pranavk-2003 commented:
> 1. The lesson numbers are incorrect, but I am not sure how important this is
> 2. The lesson about the PID regulator is quite lengthy, but it might be acceptable as a temporary lesson.
> 
> @spivakdm what do you think about it?

i'll recheck and rework the the lesson numbers once again.Controller section will have more 2 lessons ill add them soon

spivakdm commented:
@ShulcN 
@pranavk-2003 

Could you double check the numbers of the lessons. It is hard for me to tell, from the frontend it seems that there are 12 assignments but Pranav says that drawing is assignment number 0 . 

PID lesson is fine for now, we can later discuss how to do it step by step in multiple assignments.

pranavk-2003 commented:
@ShulcN I added more content to it and also resolved the long lesson of PID into 2 lessons.Please review it

spivakdm commented:
@pranavk-2003 
1. there is no verifications , only lesson texts. we were only able to find verifications for module 6. Can you add them to PR, what is your plan with other modules and verifications. 

2. remove LED matrix lesson from PR  we are not able to detect it yet.

3. you have 2 lessons called PID.md in different modules

pranavk-2003 commented:
- The verification are located in the Issues tab where i made a new issue for them and made it a TASK.They only contained the verifications of module 6 that is correct.
- As we removed the LED Matrix assignment , it leaves us with module 7(basic line follower and line sensor intro) and module 9(P and PI , PID)I do have those verification with me. i can include those in a seperate PR as it would be easier to check.
- Removed the LED matrix.
- Also i saw the PID being repeated twice, that happened as a result of branch resolve conflict, i will fixed that.

spivakdm commented:
@pranavk-2003 I think line follower should be probably 1st assignment of module 6
module 8 will be module 7
and first assignment in module 7 will be relay controller, then PI, then PID

pranavk-2003 commented:
Pushed the modified assignment for 6.1 ,6.2 is the same the verification would just change , not the content.The verification shall shall be realistic and have loosen conditions to be passed the assignment.Although did not get the chance to verify it on the robot for 6.2.Below is verification and robot code for 6.1

Robot Code :
```
#include <Octoliner.h>
#include <lineRobot.h>

// I2C Address (default 42)
Octoliner octoliner(42);

void setup() {
    octoliner.begin();
    octoliner.setSensitivity(245);
}

void loop() {
    // Read values from two sensors
    int value_1 = octoliner.analogRead(1);
    int value_2 = octoliner.analogRead(2);
    int value_3 = octoliner.analogRead(3);
    int value_4 = octoliner.analogRead(4);
    int value_5 = octoliner.analogRead(5);
    int value_6 = octoliner.analogRead(6);
    int value_7 = octoliner.analogRead(7);

    if(value_1>300){
        printMQTT("SENSOR 1 ON LINE");
    }
    if(value_2>300){
        printMQTT("SENSOR 2 ON LINE");
    }
    if(value_3>300){
        printMQTT("SENSOR 3 ON LINE");
    }
    if(value_4>300){
        printMQTT("SENSOR 4 ON LINE");
    }
    if(value_5>300){
        printMQTT("SENSOR 5 ON LINE");
    }
    if(value_6>300){
        printMQTT("SENSOR 6 ON LINE");
    }
    if(value_7>300){
        printMQTT("SENSOR 7 ON LINE");
    }
    else{
        printMQTT("NO SENSOR ON THE LINE");
    }
}
```
Start Point  : X : 28 ,Y : 58

Verification code :

```
```

pranavk-2003 commented:
@ShulcN These are the Updated codes and verifications for module 6.Please use these to test.

### Assignment 6.1

Starting point X : 40 ,Y : 50

```
def task_test(robot, image, td):
    """
    Verification function that checks if at least one sensor (0-7) is detected as ON THE LINE.
    """
    result = {
        "success": True,  # Start with success assumption
        "description": "Waiting for sensor messages...",
        "score": 100
    }
    
    # Use a consistent verification message until the end
    text = "Verifying..."
    
    # Draw robot info on image
    image = robot.draw_info(image)
    
    # Initialize test data if not already set
    if not td:
        td = {
            "end_time": time.time() + 10,  # 10 seconds verification time
            "data": {
                "sensors_on_line": set(),  # Track which sensors are detected as on the line
                "messages": [],
                "start_time": time.time(),
                "result_displayed": False  # Flag to track if final result is displayed
            }
        }
    
    # Get message from robot
    msg = robot.get_msg()
    if msg is not None:
        # Store all messages
        td["data"]["messages"].append(msg)
        
        # Check for any sensor ON LINE messages (0-7)
        for sensor_num in range(8):
            if f"SENSOR {sensor_num} ON LINE" in msg.upper():
                td["data"]["sensors_on_line"].add(sensor_num)
    
    # Check if time has expired
    if time.time() > td["end_time"]:
        # Set the final result text only once when time expires
        if not td["data"]["result_displayed"]:
            td["data"]["result_displayed"] = True
            
            if td["data"]["sensors_on_line"]:  # If any sensor is on the line
                result["success"] = True
                detected_sensors = ", ".join(str(s) for s in sorted(td["data"]["sensors_on_line"]))
                result["description"] = f"Assignment passed! Detected sensors on line: {detected_sensors}"
                text = "Verification successful!"
            else:
                result["success"] = False
                
                if not td["data"]["messages"]:
                    result["description"] = "No messages received from robot."
                else:
                    result["description"] = "No sensor was detected as ON LINE."
                
                text = "Verification failed."
                result["score"] = 0
    
    return image, td, text, result
```

Robot Code:

```
#include <Octoliner.h>
#include <lineRobot.h>

// I2C Address (default 42)
Octoliner octoliner(42);

void setup() {
    octoliner.begin();
    octoliner.setSensitivity(245);
}

void loop() {
    // Read values from two sensors
    int value_1 = octoliner.analogRead(1);
    int value_2 = octoliner.analogRead(2);
    int value_3 = octoliner.analogRead(3);
    int value_4 = octoliner.analogRead(4);
    int value_5 = octoliner.analogRead(5);
    int value_6 = octoliner.analogRead(6);
    int value_7 = octoliner.analogRead(7);

    if(value_1>300){
        printMQTT("SENSOR 1 ON LINE");
    }
    if(value_2>300){
        printMQTT("SENSOR 2 ON LINE");
    }
    if(value_3>300){
        printMQTT("SENSOR 3 ON LINE");
    }
    if(value_4>300){
        printMQTT("SENSOR 4 ON LINE");
    }
    if(value_5>300){
        printMQTT("SENSOR 5 ON LINE");
    }
    if(value_6>300){
        printMQTT("SENSOR 6 ON LINE");
    }
    if(value_7>300){
        printMQTT("SENSOR 7 ON LINE");
    }
    else{
        printMQTT("NO SENSOR ON THE LINE");
    }
}
```

### Assignment 6.2

Starting point X : 40 ,Y : 50

Verification code:

```
def task_test(robot, image, td):
    """
    Robust verification function that waits for the full time period
    before making any pass/fail decision.
    """
    result = {
        "success": True,  # Start with success assumption
        "description": "Collecting sensor data...",
        "score": 100
    }
    text = "Verifying sensor values..."
    
    # Draw robot info on image
    image = robot.draw_info(image)
    
    # Initialize test data if not already set
    if not td:
        td = {
            "end_time": time.time() + 15,  # 15 seconds verification time
            "data": {
                "sensor_values": {},  # Store sensor values as they come in
                "sensors_above_threshold": set(),  # Track which sensors are above threshold
                "messages": [],
                "start_time": time.time(),
                "verification_complete": False  # Flag to ensure we only complete once
            }
        }
    
    # Get message from robot
    msg = robot.get_msg()
    if msg is not None:
        # Store message
        td["data"]["messages"].append(msg)
        
        # Parse sensor messages in format "Sensor X: Y"
        if "Sensor" in msg and ":" in msg:
            try:
                # Split by colon and extract parts
                parts = msg.split(":")
                if len(parts) == 2:
                    sensor_part = parts[0].strip()
                    value_part = parts[1].strip()
                    
                    # Extract sensor number
                    sensor_num = int(sensor_part.replace("Sensor", "").strip())
                    
                    # Extract sensor value
                    sensor_val = int(value_part)
                    
                    # Store the sensor value
                    td["data"]["sensor_values"][sensor_num] = sensor_val
                    
                    # Check if value is above threshold
                    if sensor_val > 200:
                        td["data"]["sensors_above_threshold"].add(sensor_num)
            except:
                # If parsing fails, just continue
                pass
    
    # CRITICAL: Do not change success/score until time expires
    # This prevents immediate failure
    
    # Check if time has expired - ONLY place we set success/failure
    if time.time() > td["end_time"] and not td["data"]["verification_complete"]:
        td["data"]["verification_complete"] = True
        sensors_above_threshold = len(td["data"]["sensors_above_threshold"])
        
        if sensors_above_threshold >= 2:
            result["success"] = True
            above_threshold_list = ", ".join(str(s) for s in sorted(td["data"]["sensors_above_threshold"]))
            result["description"] = f"Assignment passed! Found {sensors_above_threshold} sensors > 200: {above_threshold_list}"
            result["score"] = 100
        else:
            result["success"] = False
            if sensors_above_threshold == 0:
                result["description"] = "No sensors detected with values > 200."
            else:
                result["description"] = "Assignment Failed"
            result["score"] = 0
        
        text = "Verification complete!"
    
    return image, td, text, result
```

Robot Code:

```
#include <Octoliner.h>
#include <lineRobot.h>

// I2C Address (default 42)
Octoliner octoliner(42);
// Black threshold for detection
const int MY_BLACK_THRESHOLD = 100;

void setup() {
    octoliner.begin();
    octoliner.setSensitivity(245);  // Adjust sensitivity if needed
}

void loop() {
    for (int i = 0; i < 8; i++) {
        printMQTT("Sensor " + String(i) +": " +  octoliner.analogRead(i));
    }
    
}
```

### Assignment 6.3

Starting point X : 30 ,Y : 50

Verification code:

```
def task_test(robot, image, td):
    """
    Verification function that passes if at least 2 sensor values are above 200.
    """
    result = {
        "success": True,
        "description": "Checking for values above 200...",
        "score": 100
    }
    text = "Verifying..."
    
    # Draw robot info on image
    image = robot.draw_info(image)
    
    # Initialize test data if not already set
    if not td:
        td = {
            "end_time": time.time() + 15,
            "data": {
                "values_above_200": 0,
                "total_values": 0,  # Added for debugging
                "start_time": time.time()
            }
        }
    
    # Get message from robot
    msg = robot.get_msg()
    if msg is not None:
        # Extract any number from the message
        import re
        numbers = re.findall(r'\d+', msg)
        
        for num_str in numbers:
            try:
                value = int(num_str)
                # Check if it's a reasonable sensor value (0-1023 range)
                if 0 <= value <= 1023:
                    td["data"]["total_values"] += 1  # Count all valid values
                    if value > 200:
                        td["data"]["values_above_200"] += 1
            except:
                pass
    
    # Display debugging info on image
    cv2.putText(image, f"Total values received: {td['data']['total_values']}", 
                (20, 30), cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 1)
    
    cv2.putText(image, f"Values above 200: {td['data']['values_above_200']}", 
                (20, 60), cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 1)
    
    # Check if time has expired
    if time.time() > td["end_time"]:
        if td["data"]["values_above_200"] >= 2:
            result["success"] = True
            result["description"] = f"Assignment passed! Found {td['data']['values_above_200']} values > 200 out of {td['data']['total_values']} total values"
            result["score"] = 100
        else:
            result["success"] = False
            result["description"] = "Assignment Failed!"
            result["score"] = 0
        
        text = "Verification complete!"
    
    return image, td, text, result
```

Robot Code:

```
#include <Octoliner.h>
#include <lineRobot.h>

// I2C Address (default 42)
Octoliner octoliner(42);
// Black threshold for detection
const int MY_BLACK_THRESHOLD = 100;

void setup() {
    octoliner.begin();
    octoliner.setSensitivity(245);  // Adjust sensitivity if needed
    int first_measurement[8]={};
    int second_measurement[8]={};
    int third_measurement[8]={};
    for(int j=0;j<=7;j++){
    first_measurement[j] = octoliner.analogRead(j);
    }
    robot.moveForwardDistance(10);
    delay(1000);
    for(int k=0;k<=7;k++){
    second_measurement[k] = octoliner.analogRead(k);
    }
    robot.moveForwardDistance(10);
    delay(1000);
    for(int l=0;l<=7;l++){
    third_measurement[l] = octoliner.analogRead(l);
    }
    robot.moveForwardDistance(10);
    delay(1000);
    for(int i=0;i<=7;i++){
        delay(100);
        printMQTT(first_measurement[i]);
        delay(100);
        printMQTT(second_measurement[i]);
        delay(100);
        printMQTT(third_measurement[i]);
        delay(100);
        
    }
}

void loop() {
    
}
```

ShulcN commented:
# 6.1

1. I had a problem with the start position; I've changed it to 35, 50.
2. I also recommend adding a delay to sending messages

# 6.2 

Same comments as for task 6.1.

# 6.3
Everything works fine.
